src/symmetric_crypto/aes_256_gcm_pure/mod.rs

```diff
@@ -3,73 +3,96 @@
 //!
 //! It will use the AES native interface on the CPU if available.
 use crate::{
-    symmetric_crypto::{nonce::NonceTrait, SymmetricCrypto},
+    symmetric_crypto::{nonce::NonceTrait, Dem, SymKey},
     CryptoCoreError,
 };
-use aes::cipher::Unsigned;
 use aes_gcm::{
-    aead::{Aead, NewAead, Payload},
-    AeadInPlace, Aes256Gcm,
+    aead::{Aead, Payload},
+    aes::cipher::generic_array::GenericArray,
+    AeadInPlace, Aes256Gcm, KeyInit,
 };
-use generic_array::{typenum, GenericArray};
-use std::fmt::Display;
+use rand_core::{CryptoRng, RngCore};
 
-pub mod dem;
+use super::{key::Key, nonce::Nonce};
 
 /// Use a 256-bit AES key
-pub type KeyLength = typenum::U32;
+const KEY_LENGTH: usize = 32;
 
 /// Use a 96-bit nonce
-pub const NONCE_LENGTH: usize = 12;
+const NONCE_LENGTH: usize = 12;
 
 /// Use a 128-bit MAC tag
-pub const MAC_LENGTH: usize = 16;
+const MAC_LENGTH: usize = 16;
 
-pub type Key = crate::symmetric_crypto::key::Key<KeyLength>;
-
-pub type Nonce = crate::symmetric_crypto::nonce::Nonce<NONCE_LENGTH>;
+/// A 96-bit nonce restricts the plaintext size to 4096 bytes
+const MAX_PLAINTEXT_LENGTH: usize = 4096;
```

Comment:
Is it possible to compute `4096` from `NONCE_LENGTH` with a simple formula if they are linked?

Reply:
Sadly, we cannot. We would need to write something like that:
``` rust
const MAX_PLAINTEXT_LENGTH: usize = 2.pow(NONCE_LENGTH);
```

But this does not work since we want to have a constant.

Reply:
It seems to work in Rust https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b0f8a45bbd2793619b5a524fce4194c7

Reply:
Indeed, it seems to work, my bad !

Reply:
Quick note: i think we'd rather to use `1 << NONCE_LENGTH` instead of `pow`.

Reply:
Actually, it does not depend on the once length. Apparently AES GCM uses a fixed 32-bit length counter and thus the maximum size of a plaintext is `2 ^ 39 - 256` bits.

I was using the `pow` to stick to the formula, but I could write it like that:
``` rust
const MAX_PLAINTEXT_LENGTH: u64 = (1 << 39 - 32) >> 3;
```
or like that:
``` rust
const MAX_PLAINTEXT_LENGTH: u64 = 68_719_476_704;
```
Which way do you think is the best ?

Reply:
Proposal:
```rust
const MAX_PLAINTEXT_LENGTH: u64 = 68_719_476_704; // 2 ^ 39 - 256 bits
```